event-publisher/src/event_layer.rs

```diff
@@ -0,0 +1,66 @@
+use mbus_api::event::EventMessage;
+use std::fmt::Debug;
+use tokio::sync::mpsc::{error::TrySendError, Sender};
+use tracing::field::{Field, Visit};
+use tracing_subscriber::{layer::Context, Layer};
+
+/// Custom layer for recording the events for mbus.
+pub struct EventLayer {
+    sender: Sender<EventMessage>,
+}
+
+impl EventLayer {
+    // Creates the new EventLayer. The sender handle for the buffer is required to store the events
+    // in buffer as the events are recorded.
+    pub fn new(sender: Sender<EventMessage>) -> Self {
+        Self { sender }
+    }
+}
+
+// Notifies the EventLayer that an event has occurred.
+impl<S> Layer<S> for EventLayer
+where
+    S: tracing::Subscriber,
+{
+    // Records an event, gets the EventMessage from the event and sends it to the buffer.
+    fn on_event(&self, event: &tracing::Event<'_>, _ctx: Context<'_, S>) {
+        let mut visitor = EventVisitor::default();
+
+        event.record(&mut visitor);
+
+        let output = visitor.0;
+
+        if let Err(err) = self.sender.try_send(output) {
+            match err {
+                TrySendError::Full(_) => {
+                    tracing::warn!("Buffer is full");
```

Comment:
same comment as on other repo

Reply:
Should I add the timer in the next PR. Currently when I tried adding the timer, the on_event method takes &self. I tried with interior mutability, but I am getting some errors.

Reply:
Yes sure, make it trace! for now then.

Reply:
This is resolved but it's still a warn and not a trace

Reply:
Changed it to trace now.